--- iscsitarget-0.4.15/kernel/config.c~	2008-12-27 22:14:44.000000000 +0100
+++ iscsitarget-0.4.15/kernel/config.c	2009-07-30 10:40:07.000000000 +0200
@@ -5,6 +5,7 @@
  */
 
 #include <linux/proc_fs.h>
+#include <linux/version.h>
 
 #include "iscsi.h"
 #include "iscsi_dbg.h"
@@ -43,7 +44,9 @@ int iet_procfs_init(void)
 	if (!(proc_iet_dir = proc_mkdir("iet", init_net.proc_net)))
 		goto err;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,31)
 	proc_iet_dir->owner = THIS_MODULE;
+#endif
 
 	for (i = 0; i < ARRAY_SIZE(iet_proc_entries); i++) {
 		ent = create_proc_entry(iet_proc_entries[i].name, 0, proc_iet_dir);
