diff -up iscsitarget-0.4.15/kernel/block-io.c~ iscsitarget-0.4.15/kernel/block-io.c
--- iscsitarget-0.4.15/kernel/block-io.c~	2008-09-23 09:20:55.000000000 +0200
+++ iscsitarget-0.4.15/kernel/block-io.c	2008-09-23 09:20:55.000000000 +0200
@@ -9,6 +9,7 @@
  * This code is licenced under the GPL.
  */
 
+#include <linux/types.h>
 #include <linux/parser.h>
 #include <linux/blkdev.h>
 #include <linux/buffer_head.h>
